[WIP][DO NOT MERGE][13.0] Add sale_warehouse_calendar

mmequignon opened the pull request:
:warning: 
Work in progress
Needs fixing
Do not merge

Computes sale order lead times with respect to to Warehouse calendar

mmequignon committed:
[13.0] Add sale_warehouse_calendar

sebalix commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,37 @@
+            td_customer_lead = timedelta(days=customer_lead)
+            # remove customer lead added to order_date in sale_stock
+            res -= td_customer_lead
+            workload = 1 + customer_lead + self.company_id.security_lead
```
```suggestion
            workload = 1 + customer_lead - self.company_id.security_lead
```

sebalix commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,37 @@
+        return res
+
+    def _prepare_procurement_values(self, group_id=False):
+        res = super()._prepare_procurement_values()
```
```suggestion
        res = super()._prepare_procurement_values(group_id)
```

sebalix commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,37 @@
+            date_planned = res.get("date_planned")
+            if date_planned:
+                customer_lead = self.customer_lead or 0.0
+                workload = 1 + customer_lead - self.company_id.security_lead
```
to extract in a method

simahawk commented on sale_warehouse_calendar/tests/test_sale_order_dates.py:
```diff
@@ -0,0 +1,124 @@
+        cls.product = cls.env.ref("product.product_product_3")
+
+    @classmethod
+    def setUpClass(cls):
```
move this to the top pls

simahawk commented on sale_warehouse_calendar/tests/test_sale_order_dates.py:
```diff
@@ -0,0 +1,124 @@
+        #   has not postponed the delivery
+        # - friday is compatible with the warehouse calendar.
+        # - delivery should be done on satursday
+        order = self._create_order(datetime.now())
```
if you use datetime.now you are not freezing anything... :thinking:

mmequignon replied:
https://github.com/spulec/freezegun#context-manager
I think it's precisely the usecase, no ?

mmequignon replied:
![image](https://user-images.githubusercontent.com/13200247/133081183-4c3d37fd-60b7-43ad-9bde-3ec1bf21e48c.png)

simahawk replied:
the point is that you are freezing the date w/ a dynamic date. You should pass a specific date to satisfy your test IMO.

mmequignon replied:
I'm using `freeze_time` as a decorator for the test methods, therefore, when calling `_create_order(datetime.now())`, the date is already frozen.
Probably not very efficient, I'm not sure I have to freeze the date in `_create_order`, but I'm pretty sure the date is frozen well.

sebalix replied:
Why calling `freeze_time` in `_create_order` method, can't we rely only on the freeze done by the decorator on the test method?

simahawk commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+    _inherit = "sale.order.line"
+
+    def _get_workload(self, customer_lead):
+        return 1 + customer_lead - self.company_id.security_lead
```
any comment regarding adding `1`?

jbaudoux replied:
This is to convert a delay into numbers of days.
https://en.wikipedia.org/wiki/Off-by-one_error#Fencepost_error

sebalix replied:
So a comment is welcome

mmequignon replied:
It has been done already…

simahawk commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+        if calendar:
+            customer_lead = self.customer_lead or 0.0
+            td_customer_lead = timedelta(days=customer_lead)
+            # remove customer lead added to order_date in sale_stock
```
are we always sure this time is included?

simahawk replied:
yes it is the core behavior... :smile:

sebalix commented on .travis.yml:
```diff
@@ -27,10 +27,10 @@ jobs:
     - stage: test
       env:
-        - TESTS="1" ODOO_REPO="OCA/OCB"
+        - TESTS="1" ODOO_REPO="OCA/OCB" EXCLUDE="sale_warehouse_calendar"
```
Can you add a comment about the conflicting module?

mmequignon replied:
done in the doc.

simahawk replied:
are you sure you are running your tests w/ these settings?

sebalix reviewed (approved):
Pre-approving, one small comment

jbaudoux commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,47 @@
+        if date_planned and calendar:
+            customer_lead = self.customer_lead or 0.0
+            workload = self._get_workload(customer_lead)
+            res["date_planned"] = calendar.plan_days(workload, date_planned)
```
Here you need to subtract the workload in delay from the date_planned before counting the days.
```suggestion
            res["date_planned"] = calendar.plan_days(workload, date_planned - timedelta(workload - 1))
```

Example:

| | |
| --- | --- |
| customer lead | 3 |
| security | 1 |
| **without this module:** |   |
| order date | fri |
| expected date | mon |
| workload in days | 3 |
| work days | fri/sat/sun |
| planned date | sun |
| **with this module:** | |
| WH calendar | mon to fri |
| work days | fri/mon/tue |
| expected date | wed |
| planned date | tue |

jbaudoux commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,47 @@
+    _inherit = "sale.order.line"
+
+    def _get_workload(self, customer_lead):
+        # Since the workload is a number of days, it cannot be 0
```
I would add a docstring instead of a comment saying that you compute a workload in number of days

Explaining that you convert a delay into a numbers of days (hence the +1) is more understandable than "it cannot be 0" ;-)

jbaudoux commented on sale_warehouse_calendar/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,5 @@
+Computes sale order lead times with respect to to Warehouse calendar
```
```suggestion
Compute the sales order expected date and delivery order planned date respecting the warehouse calendar
```

jbaudoux commented on sale_warehouse_calendar/tests/test_sale_order_dates.py:
```diff
@@ -0,0 +1,124 @@
+CUTOFF_TIME = 8.0  # cutoff time will be set at 8 a.m.
+
+FRIDAY = "2021-08-13"
+SATURSDAY = "2021-08-14"
```
SATURDAY

jbaudoux commented on sale_warehouse_calendar/tests/test_sale_order_dates.py:
```diff
@@ -0,0 +1,124 @@
+        # - preparation is friday since sale_cutoff_time_delivery,
+        #   has not postponed the delivery
+        # - friday is compatible with the warehouse calendar.
+        # - delivery should be done on satursday
```
saturday

jbaudoux commented on sale_warehouse_calendar/tests/test_sale_order_dates.py:
```diff
@@ -0,0 +1,124 @@
+            )
+            return order
+
+    @freeze_time(FRIDAY_BEFORE_CUTOFF)
```
You should make an order on Saturday before cutoff and you should see this test failing without the fix I mentioned

jbaudoux commented on sale_warehouse_calendar/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+
+{
+    "name": "Sale Warehouse Calendar",
+    "summary": "Computes sale order lead times with respect to to Warehouse calendar",
```
```suggestion
    "summary": "Compute sales order expected date and delivery order planned date respecting the warehouse calendar",
```

jbaudoux reviewed (changes requested):
There is likely an error in the computation of the date planned

jbaudoux reviewed (approved):
Thanks for the comments. Code is very understandable

simahawk commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -0,0 +1,59 @@
+        date_planned = res.get("date_planned")
+        calendar = self.order_id.warehouse_id.calendar_id
+        if date_planned and calendar:
+            # customer_lead is security_lead + workload, as explained on the field
+            customer_lead = self.customer_lead or 0.0
+            security_lead = self.company_id.security_lead or 0.0
+            workload = customer_lead - security_lead
+            # plan_days() expect a number of days instead of a delay
+            workload_days = self._delay_to_days(workload)
```
The part for the workload looks the same as the method above: can't you move this to a reusable method?

simahawk reviewed (changes requested):
check the tests pls

mmequignon committed:
TO REMOVE: add temporary calendar field on warehouse

DDMRP is broken when a calendar is set on the warehouse.
We need to be able to set a calendar whithout breaking DDMRP.
Remove this commit when DDMRP is fixed.

mmequignon commented:
Do not merge this (see the last commit description)

mmequignon committed:
fixup! [13.0] Add sale_warehouse_calendar

jbaudoux commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -41,6 +41,11 @@ def _sale_warehouse_calendar_expected_date(self, expected_date):
             expected_date += td_security_lead
+            # This should be the delivery date.
+            # It is up to the carrier to determine the hour with the customer
+            expected_date = expected_date.replace(
```
This is not TZ compliant and won't work for countries in UTC-X.
I would not change the hours here. If this is necessary in a later calendar computation, it is there that it should be fixed.

jbaudoux commented on sale_warehouse_calendar/models/sale_order_line.py:
```diff
@@ -58,7 +63,10 @@ def _sale_warehouse_calendar_prepare_procurement_values(self, res):
+            date_planned = calendar.plan_days(
                 workload_days, date_planned, compute_leaves=True
             )
+            res["date_planned"] = date_planned.replace(
```
Same remark as above

jbaudoux reviewed (changes requested):
This last commit is not fixing the hours issue in the right way.
If calendar plan_day is hours aware, then it should be fixed just before the call to that method.
